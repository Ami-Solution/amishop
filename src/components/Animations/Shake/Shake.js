import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes, css } from 'styled-components';

const shakeAnim = keyframes`
  10%, 90% {transform: translate3d(-1px, 0, 0);}
  20%, 80% {transform: translate3d(2px, 0, 0);}
  30%, 50%, 70% {transform: translate3d(-4px, 0, 0);}
  40%, 60% {transform: translate3d(4px, 0, 0);}
`;

const ShakeWrapper = styled.div`
  transform: translate3d(0, 0, 0);
  backface-visibility: hidden;
  perspective: 1000px;

  ${({ shake }) =>
    shake &&
    css`
      animation: ${shakeAnim} 0.82s cubic-bezier(0.36, 0.07, 0.19, 0.97) both;
    `};
`;

/**
 * Shake component Animation
 * @extends PureComponent
 */

class Shake extends PureComponent {
  static propTypes = {
    toggle: PropTypes.number,
    children: PropTypes.oneOfType([PropTypes.array, PropTypes.element])
      .isRequired,
  };

  static defaultProps = {
    toggle: 0,
  };

  state = {
    shake: false,
    toggle: 0,
  };

  componentWillReceiveProps = ({ toggle: propToggle }) => {
    const { toggle: stateToggle } = this.state;

    if (propToggle > stateToggle) {
      this.setState({ shake: true, toggle: propToggle });
      setTimeout(() => this.setState({ shake: false }), 850);
    }
  };

  render = () => {
    const { children } = this.props;
    const { shake } = this.state;

    return <ShakeWrapper shake={shake}>{children}</ShakeWrapper>;
  };
}

export default Shake;
