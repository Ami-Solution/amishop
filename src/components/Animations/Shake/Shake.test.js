/* global describe it expect */
import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Shake from './index';

configure({ adapter: new Adapter() });

describe('Component Shake', () => {
  const child = <p>Child</p>;
  const component = renderer.create(<Shake toggle={5}>{child}</Shake>);
  const tree = component.toJSON();

  it('should be render without crash', () => {
    expect(tree).toMatchSnapshot();
  });
});
