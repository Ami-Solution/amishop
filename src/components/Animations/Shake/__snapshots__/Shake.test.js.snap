// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Shake should be render without crash 1`] = `
<div
  className="sc-bdVaJa elOHmT"
>
  <p>
    Child
  </p>
</div>
`;
